### PR TITLE
[REFACTOR] Replace TensorObj and TensorValue with NDArray

### PR DESCRIPTION
**Pull request by zhiics**
This PR removes the TensorObj in VM and TensorValue in interpreter. Instead, we use NDArray directly since it has been migrated to the unified object protocol, #4599 

Update: NodeBase is also replaced by Object in this PR.

### CODE REVIEW

**Comment by zhiics** [comment]
cc @tqchen @wweic @icemelon9 @jroesch

File: python/tvm/relay/backend/interpreter.py [hunk 1]
```diff
@@ -23,27 +23,13 @@
...
-        """Convert a Python scalar to a Relay scalar."""
-        return TensorValue(const(value, dtype).data)
-
-    def to_vm(self):
-        return _vm._ValueToVM(self)
-
 
 @register_relay_node
-class TupleValue(Value):
+class TupleValue(NodeBase):
```
Comment (jroesch) [inline python/tvm/relay/backend/interpreter.py]: Shouldn't this be Object? or the equivalent Object base class, seems wrong to inherit from node. cc @tqchen

Comment (zhiics) [reply python/tvm/relay/backend/interpreter.py]: I noticed this as well. It is from here:

https://github.com/apache/incubator-tvm/blob/bc0274d307226408c69226cf922dd916d773e265/python/tvm/_ffi/node.py#L33

and used as the base by `Value` before so I just kept it.

File: include/tvm/relay/interpreter.h [hunk 2]
```diff
@@ -65,39 +61,21 @@ class Value;
...
-
-  using ContainerType = ValueNode;
-};
-
 /*! \brief A Relay closure, i.e a scope and a function. */
 class Closure;
 
 /*! \brief The container type of Closures. */
-class ClosureNode : public ValueNode {
+class ClosureNode : public Object {
```
Comment (wweic) [inline include/tvm/relay/interpreter.h]:
```suggestion
class ClosureObj : public Object {
```

File: include/tvm/relay/interpreter.h [hunk 3]
```diff
@@ -111,22 +89,22 @@ class ClosureNode : public ValueNode {
...
-  TVM_DEFINE_OBJECT_REF_METHODS(Closure, Value, ClosureNode);
+  TVM_DEFINE_OBJECT_REF_METHODS(Closure, ObjectRef, ClosureNode);
 };
 
 /*! \brief A Relay Recursive Closure. A closure that has a name. */
 class RecClosure;
 
 /*! \brief The container type of RecClosure. */
-class RecClosureNode : public ValueNode {
+class RecClosureNode : public Object {
```
Comment (wweic) [inline include/tvm/relay/interpreter.h]:
```suggestion
class RecClosureObj : public Object {
```

File: include/tvm/relay/interpreter.h [hunk 4]
```diff
@@ -143,89 +121,66 @@ class RecClosureNode : public ValueNode {
...
+  TVM_DEFINE_OBJECT_REF_METHODS(RecClosure, ObjectRef, RecClosureNode);
 };
 
 /*! \brief A tuple value. */
 class TupleValue;
 
 /*! \brief Tuple (x, ... y). */
-struct TupleValueNode : ValueNode {
-  tvm::Array<Value> fields;
+struct TupleValueNode : Object {
```
Comment (wweic) [inline include/tvm/relay/interpreter.h]:
```suggestion
struct TupleValueObj : Object {
```

**Comment by zhiics** [comment]
@wweic I am trying to only refactor tensors in the PR. The other ones may need to be either unified between VM and interpreter or changed. We will need another PR for this.

**Comment by wweic** [comment]
> @wweic I am trying to only refactor tensors in the PR. The other ones may need to be either unified between VM and interpreter or changed. We will need another PR for this.

Make senses. I expect it will need more changes.

File: python/tvm/relay/backend/interpreter.py (hunk 1 above)
Comment (tqchen) [reply python/tvm/relay/backend/interpreter.py]: There is no Node anymore in the c++ side, see https://github.com/apache/incubator-tvm/pull/4603 and https://github.com/apache/incubator-tvm/issues/4647

So NodeBase was a legacy item that we might need to upgrade on the python side.

Comment (tqchen) [reply python/tvm/relay/backend/interpreter.py]: Let us inherit from Object for now, and have a subsequent PR to cleanup the python side.

Also seems we might be able to use Array(or ADT, see discussion here https://discuss.tvm.ai/t/discuss-runtime-array-containers-array-adt-string/4582) for this after @wweic 's PR lands

Comment (zhiics) [reply python/tvm/relay/backend/interpreter.py]: Yes, I tried it, but but looks to me that we have to merge nodebase to object first before I can use object directly as there are fields not implemented in object, like getattr, etc. Should I use nodebase first then we can refactor them together?

Comment (zhiics) [reply python/tvm/relay/backend/interpreter.py]: I just added a commit to replace NodeBase with Object

**Comment by zhiics** [comment]
@jroesch can you take another look? Should we merge?

**Comment by tqchen** [comment]
@zhiics please do a rebase

**Comment by zhiics** [comment]
@tqchen Done.

**Comment by tqchen** [comment]
still need to wait for @jroesch 's approval

**Pull request merged**